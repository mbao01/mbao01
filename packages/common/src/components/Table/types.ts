import { VariantProps } from "../../libs";
import { getTableClasses } from "./constants";

export type TableProps = React.HTMLAttributes<HTMLTableElement> &
  VariantProps<typeof getTableClasses>;

export type TableHeaderProps = React.HTMLAttributes<HTMLTableSectionElement>;

export type TableBodyProps = React.HTMLAttributes<HTMLTableSectionElement>;

export type TableFooterProps = React.HTMLAttributes<HTMLTableSectionElement>;

export type TableRowProps = React.HTMLAttributes<HTMLTableRowElement>;

export type TableHeadProps = React.ThHTMLAttributes<HTMLTableCellElement>;

export type TableCellProps = React.TdHTMLAttributes<HTMLTableCellElement>;

export type TableCaptionProps = React.HTMLAttributes<HTMLTableCaptionElement>;
