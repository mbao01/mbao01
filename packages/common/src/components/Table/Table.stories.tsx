import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { Table } from "./Table";
import { type TableProps } from "./types";

const items = [
  {
    invoice: "INV001",
    status: "Paid",
    method: "Credit Card",
    amount: "$250.00",
  },
  {
    invoice: "INV002",
    status: "Unpaid",
    method: "Direct Debit",
    amount: "$250.00",
  },
  {
    invoice: "INV003",
    status: "Paid",
    method: "Credit Card",
    amount: "$250.00",
  },
  {
    invoice: "INV004",
    status: "Paid",
    method: "Direct Debit",
    amount: "$250.00",
  },
  {
    invoice: "INV005",
    status: "Pending",
    method: "Debit Card",
    amount: "$250.00",
  },
  {
    invoice: "INV006",
    status: "Paid",
    method: "Credit Card",
    amount: "$250.00",
  },
  {
    invoice: "INV007",
    status: "Paid",
    method: "Credit Card",
    amount: "$250.00",
  },
];

const withTable = (_: StoryFn, context: StoryContext<TableProps>) => {
  return (
    <Table {...context.args}>
      <Table.Caption>A list of your recent invoices.</Table.Caption>
      <Table.Header>
        <Table.Row>
          <Table.Head className="w-[100px]">Invoice</Table.Head>
          <Table.Head>Status</Table.Head>
          <Table.Head>Method</Table.Head>
          <Table.Head className="text-right">Amount</Table.Head>
        </Table.Row>
      </Table.Header>
      <Table.Body>
        {items.map((item) => (
          <Table.Row key={item.invoice}>
            <Table.Cell className="font-medium">{item.invoice}</Table.Cell>
            <Table.Cell>{item.status}</Table.Cell>
            <Table.Cell>{item.method}</Table.Cell>
            <Table.Cell className="text-right">{item.amount}</Table.Cell>
          </Table.Row>
        ))}
      </Table.Body>
      <Table.Footer>
        <Table.Row>
          <Table.Cell colSpan={3}>Total</Table.Cell>
          <Table.Cell className="text-right">$2,500.00</Table.Cell>
        </Table.Row>
      </Table.Footer>
    </Table>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  component: Table,
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withTable],
} satisfies Meta<typeof Table>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};
