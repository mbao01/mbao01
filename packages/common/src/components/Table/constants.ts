import { cva } from "../../libs";

export const getTableClasses = cva("table", {
  variants: {
    caption: {
      top: "caption-top",
      bottom: "caption-bottom",
    },
    zebra: {
      true: "table-zebra",
    },
    pinRow: {
      true: "table-pin-rows",
    },
    pinColumn: {
      true: "table-pin-cols",
    },
    size: {
      xs: "table-xs",
      sm: "table-sm",
      md: "table-md",
      lg: "table-lg",
    },
  },
});

export const getTableHeaderClasses = cva("[&_tr]:border-b");
