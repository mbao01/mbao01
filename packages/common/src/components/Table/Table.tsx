import { forwardRef } from "react";
import type {
  TableBodyProps,
  TableCaptionProps,
  TableCellProps,
  TableFooterProps,
  TableHeaderProps,
  TableHeadProps,
  TableProps,
  TableRowProps,
} from "./types";
import { cn } from "../../utilities";
import { getTableClasses, getTableHeaderClasses } from "./constants";

const Table = ({ className, caption, zebra, size, pinRow, pinColumn, ...props }: TableProps) => (
  <div className="relative w-full overflow-auto">
    <table
      className={cn(getTableClasses({ caption, zebra, size, pinRow, pinColumn }), className)}
      {...props}
    />
  </div>
);
Table.displayName = "Table";

const TableHeader = forwardRef<HTMLTableSectionElement, TableHeaderProps>(
  ({ className, ...props }, ref) => (
    <thead ref={ref} className={cn(getTableHeaderClasses(), className)} {...props} />
  )
);
TableHeader.displayName = "TableHeader";

const TableBody = forwardRef<HTMLTableSectionElement, TableBodyProps>((props, ref) => (
  <tbody ref={ref} {...props} />
));
TableBody.displayName = "TableBody";

const TableFooter = forwardRef<HTMLTableSectionElement, TableFooterProps>((props, ref) => (
  <tfoot ref={ref} {...props} />
));
TableFooter.displayName = "TableFooter";

const TableRow = forwardRef<HTMLTableRowElement, TableRowProps>((props, ref) => (
  <tr ref={ref} {...props} />
));
TableRow.displayName = "TableRow";

const TableHead = forwardRef<HTMLTableCellElement, TableHeadProps>((props, ref) => (
  <th ref={ref} {...props} />
));
TableHead.displayName = "TableHead";

const TableCell = forwardRef<HTMLTableCellElement, TableCellProps>((props, ref) => (
  <td ref={ref} {...props} />
));
TableCell.displayName = "TableCell";

const TableCaption = forwardRef<HTMLTableCaptionElement, TableCaptionProps>((props, ref) => (
  <caption ref={ref} {...props} />
));
TableCaption.displayName = "TableCaption";

Table.Header = TableHeader;
Table.Body = TableBody;
Table.Footer = TableFooter;
Table.Head = TableHead;
Table.Row = TableRow;
Table.Cell = TableCell;
Table.Caption = TableCaption;

export { Table };
