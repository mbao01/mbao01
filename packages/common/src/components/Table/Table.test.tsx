import { render, screen } from "@testing-library/react";
import { describe, expect, it } from "vitest";
import { Table } from "./Table";

describe("Table", () => {
  const items = [
    {
      invoice: "INV001",
      status: "Paid",
      method: "Credit Card",
      amount: "$250.00",
    },
    {
      invoice: "INV002",
      status: "Unpaid",
      method: "Direct Debit",
      amount: "$250.00",
    },
    {
      invoice: "INV003",
      status: "Paid",
      method: "Credit Card",
      amount: "$250.00",
    },
    {
      invoice: "INV004",
      status: "Paid",
      method: "Direct Debit",
      amount: "$250.00",
    },
    {
      invoice: "INV005",
      status: "Pending",
      method: "Debit Card",
      amount: "$250.00",
    },
    {
      invoice: "INV006",
      status: "Paid",
      method: "Credit Card",
      amount: "$250.00",
    },
    {
      invoice: "INV007",
      status: "Paid",
      method: "Credit Card",
      amount: "$250.00",
    },
  ];

  const renderTable = () => {
    return render(
      <Table>
        <Table.Caption>A list of your recent invoices.</Table.Caption>
        <Table.Header>
          <Table.Row>
            <Table.Head className="w-[100px]">Invoice</Table.Head>
            <Table.Head>Status</Table.Head>
            <Table.Head>Method</Table.Head>
            <Table.Head className="text-right">Amount</Table.Head>
          </Table.Row>
        </Table.Header>
        <Table.Body>
          {items.map((item) => (
            <Table.Row key={item.invoice}>
              <Table.Cell className="font-medium">{item.invoice}</Table.Cell>
              <Table.Cell>{item.status}</Table.Cell>
              <Table.Cell>{item.method}</Table.Cell>
              <Table.Cell className="text-right">{item.amount}</Table.Cell>
            </Table.Row>
          ))}
        </Table.Body>
        <Table.Footer>
          <Table.Row>
            <Table.Cell colSpan={3}>Total</Table.Cell>
            <Table.Cell className="text-right">$2,500.00</Table.Cell>
          </Table.Row>
        </Table.Footer>
      </Table>
    );
  };
  it("renders a table", () => {
    const { asFragment } = renderTable();

    expect(
      screen.getByRole("table", { name: "A list of your recent invoices." })
    ).toBeInTheDocument();

    items.forEach(({ invoice, status, method, amount }) => {
      expect(
        screen.getByRole("row", {
          name: `${invoice} ${status} ${method} ${amount}`,
        })
      ).toBeInTheDocument();
    });
    expect(asFragment()).toMatchSnapshot();
  });
});
