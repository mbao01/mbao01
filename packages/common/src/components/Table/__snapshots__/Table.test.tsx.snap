// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Table > renders a table 1`] = `
<DocumentFragment>
  <div
    class="relative w-full overflow-auto"
  >
    <table
      class="table"
    >
      <caption>
        A list of your recent invoices.
      </caption>
      <thead
        class="[&_tr]:border-b"
      >
        <tr>
          <th
            class="w-[100px]"
          >
            Invoice
          </th>
          <th>
            Status
          </th>
          <th>
            Method
          </th>
          <th
            class="text-right"
          >
            Amount
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="font-medium"
          >
            INV001
          </td>
          <td>
            Paid
          </td>
          <td>
            Credit Card
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV002
          </td>
          <td>
            Unpaid
          </td>
          <td>
            Direct Debit
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV003
          </td>
          <td>
            Paid
          </td>
          <td>
            Credit Card
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV004
          </td>
          <td>
            Paid
          </td>
          <td>
            Direct Debit
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV005
          </td>
          <td>
            Pending
          </td>
          <td>
            Debit Card
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV006
          </td>
          <td>
            Paid
          </td>
          <td>
            Credit Card
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
        <tr>
          <td
            class="font-medium"
          >
            INV007
          </td>
          <td>
            Paid
          </td>
          <td>
            Credit Card
          </td>
          <td
            class="text-right"
          >
            $250.00
          </td>
        </tr>
      </tbody>
      <tfoot>
        <tr>
          <td
            colspan="3"
          >
            Total
          </td>
          <td
            class="text-right"
          >
            $2,500.00
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</DocumentFragment>
`;
